File: backend/app/schemas/brainstorm.py
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict

from app.schemas.enrichment import EnrichmentStatus
from app.schemas.place import PlaceFields


class BrainstormItemBase(PlaceFields):
    pass


class BrainstormItemCreate(BrainstormItemBase):
    pass


class BrainstormItemOut(BrainstormItemBase):
    id: int
    trip_id: int
    user_id: int
    added_by: Optional[str] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class BrainstormMessageOut(BaseModel):
    id: int
    role: str
    content: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class BrainstormChatRequest(BaseModel):
    message: str


class BrainstormChatResponse(BaseModel):
    assistant_message: BrainstormMessageOut
    history: List[BrainstormMessageOut]


class BrainstormExtractResponse(BaseModel):
    items: List[BrainstormItemOut]
    enrichment: Optional[EnrichmentStatus] = None


class BrainstormBulkRequest(BaseModel):
    items: List[BrainstormItemBase]


class BrainstormPromoteRequest(BaseModel):
    item_ids: Optional[List[int]] = None


class BrainstormSeedMessage(BaseModel):
    role: str
    content: str


class BrainstormSeedRequest(BaseModel):
    messages: List[BrainstormSeedMessage]


class BrainstormSeedResponse(BaseModel):
    seeded: int


class PlanTripRequest(BaseModel):
    prompt: str
    timezone: Optional[str] = None


class PlanTripResponse(BaseModel):
    trip_name: str
    start_date: Optional[datetime] = None
    duration_days: int
    items: List[BrainstormItemBase]
    enrichment: Optional[EnrichmentStatus] = None
    user_output: str = ""

File: ios/Roammate/Models/Brainstorm.swift
import Foundation

struct BrainstormItemOut: Codable, Identifiable, Hashable {
    let id: Int
    let tripId: Int
    let userId: Int

    let title: String
    let description: String?
    let category: String?
    let placeId: String?
    let lat: Double?
    let lng: Double?
    let address: String?
    let photoUrl: String?
    let rating: Double?
    let priceLevel: Int?
    let types: [String]?
    let timeCategory: String?
    let addedBy: String?
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, title, description, category, lat, lng, address, rating, types
        case tripId = "trip_id"
        case userId = "user_id"
        case placeId = "place_id"
        case photoUrl = "photo_url"
        case priceLevel = "price_level"
        case timeCategory = "time_category"
        case addedBy = "added_by"
        case createdAt = "created_at"
    }
}

struct BrainstormMessage: Codable, Identifiable, Hashable {
    let id: Int
    let role: String
    let content: String
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, role, content
        case createdAt = "created_at"
    }
}

struct BrainstormChatRequest: Encodable {
    let message: String
}

struct BrainstormChatResponse: Codable {
    let assistantMessage: BrainstormMessage
    let history: [BrainstormMessage]

    enum CodingKeys: String, CodingKey {
        case assistantMessage = "assistant_message"
        case history
    }
}

struct BrainstormExtractResponse: Codable {
    let items: [BrainstormItemOut]
    let enrichment: JSONValue?
}

struct BrainstormPromoteRequest: Encodable {
    let itemIds: [Int]?

    enum CodingKeys: String, CodingKey {
        case itemIds = "item_ids"
    }
}
